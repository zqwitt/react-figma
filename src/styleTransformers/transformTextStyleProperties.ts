import { Color, GeometryProps, TextNodeProps } from '../types';
import { colorToRGB } from './transformColors';
import { convertFontStyle } from './converFontStyle';
import { transformDimensionMapper } from './transformDimension';
import { transformShadowToEffect } from './transformShadowToEffect';

export interface TextStyleProperties {
    color: string;
    fontFamily: string;
    fontWeight: string | number;
    fontStyle: 'normal' | 'italic';
    fontSize: number;
    textAlign: 'auto' | 'left' | 'right' | 'center' | 'justify';
    lineHeight: number | string;
    letterSpacing: number | string;
    textDecorationLine: 'none' | 'underline' | 'line-through';
    textShadowColor: Color;
    textShadowOffset: { width: number; height: number };
    textShadowRadius: number;
}

interface TextProperties extends GeometryProps, TextNodeProps {}

const textAlignMapping = {
    left: 'LEFT',
    right: 'RIGHT',
    center: 'CENTER',
    justify: 'JUSTIFIED'
};

const textDecorationLineMapping = {
    none: 'NONE',
    underline: 'UNDERLINE',
    'line-through': 'STRIKETHROUGH'
};

export const transformTextStyleProperties = (style?: Partial<TextStyleProperties>): TextProperties => {
    if (!style) {
        return {};
    }

    return {
        ...((style && style.color && { fills: [{ type: 'SOLID', color: colorToRGB(style.color) }] }) || {}),
        ...(style &&
            style.fontFamily && {
                fontName: { family: style.fontFamily, style: convertFontStyle(style.fontWeight, style.fontStyle) }
            }),
        ...(style && style.fontSize && { fontSize: style.fontSize }),
        ...(style &&
            style.textAlign &&
            textAlignMapping[style.textAlign] && { textAlignHorizontal: textAlignMapping[style.textAlign] }),
        ...(style && typeof style.lineHeight === 'number'
            ? { lineHeight: { value: style.lineHeight * 100, unit: 'PERCENT' } }
            : typeof style.lineHeight === 'string' && {
                  lineHeight: transformDimensionMapper<LineHeight, LineHeight, LineHeight>(style.lineHeight)
                      .px(value => ({ value, unit: 'PIXELS' }))
                      .percentage(value => ({ value, unit: 'PERCENT' }))
                      .auto(() => ({ unit: 'AUTO' }))
                      .value()
              }),
        ...(style &&
            (typeof style.letterSpacing === 'number' || typeof style.letterSpacing === 'string') && {
                letterSpacing: transformDimensionMapper<LetterSpacing, LetterSpacing, LetterSpacing>(
                    style.letterSpacing
                )
                    .px(value => ({ value, unit: 'PIXELS' }))
                    .percentage(value => ({ value, unit: 'PERCENT' }))
                    .value()
            }),
        ...(style &&
            style.textDecorationLine &&
            textDecorationLineMapping[style.textDecorationLine] && {
                textDecorationLine: textDecorationLineMapping[style.textDecorationLine]
            }),
        ...(style &&
            style.textShadowColor && {
                effects: [
                    transformShadowToEffect({
                        shadowColor: style.textShadowColor,
                        shadowOffset: style.textShadowOffset,
                        shadowRadius: style.textShadowRadius
                    })
                ]
            })
    };
};
