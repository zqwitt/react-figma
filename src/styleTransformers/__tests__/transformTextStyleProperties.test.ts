import { transformTextStyleProperties } from '../transformTextStyleProperties';

describe('transformTextStyleProperties', () => {
    it('textAlign auto', () => {
        const result = transformTextStyleProperties({ textAlign: 'auto' });
        expect(result).toMatchSnapshot();
    });

    it('textAlign center', () => {
        const result = transformTextStyleProperties({ textAlign: 'center' });
        expect(result).toMatchSnapshot();
    });

    it('textAlign justify', () => {
        const result = transformTextStyleProperties({ textAlign: 'justify' });
        expect(result).toMatchSnapshot();
    });

    it('lineHeight "50%"', () => {
        const result = transformTextStyleProperties({ lineHeight: '50%' });
        expect(result).toMatchSnapshot();
    });

    it('lineHeight 1.4', () => {
        const result = transformTextStyleProperties({ lineHeight: 1.4 });
        expect(result).toMatchSnapshot();
    });

    it('lineHeight auto', () => {
        const result = transformTextStyleProperties({ lineHeight: 'auto' });
        expect(result).toMatchSnapshot();
    });

    it('letterSpacing "50%"', () => {
        const result = transformTextStyleProperties({ letterSpacing: '50%' });
        expect(result).toMatchSnapshot();
    });

    it('letterSpacing 1.4', () => {
        const result = transformTextStyleProperties({ letterSpacing: 1.4 });
        expect(result).toMatchSnapshot();
    });

    it('textDecorationLine line-through', () => {
        const result = transformTextStyleProperties({ textDecorationLine: 'line-through' });
        expect(result).toMatchSnapshot();
    });

    it('text shadow', () => {
        const result = transformTextStyleProperties({
            textShadowColor: '#000000',
            textShadowRadius: 15,
            textShadowOffset: {
                width: 10,
                height: 20
            }
        });
        expect(result).toMatchSnapshot();
    });
});
