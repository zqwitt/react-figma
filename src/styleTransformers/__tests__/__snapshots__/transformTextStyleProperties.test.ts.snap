// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`transformTextStyleProperties letterSpacing "50%" 1`] = `
Object {
  "letterSpacing": Object {
    "unit": "PERCENT",
    "value": 50,
  },
}
`;

exports[`transformTextStyleProperties letterSpacing 1.4 1`] = `
Object {
  "letterSpacing": Object {
    "unit": "PIXELS",
    "value": 1.4,
  },
}
`;

exports[`transformTextStyleProperties lineHeight "50%" 1`] = `
Object {
  "lineHeight": Object {
    "unit": "PERCENT",
    "value": 50,
  },
}
`;

exports[`transformTextStyleProperties lineHeight 1.4 1`] = `
Object {
  "lineHeight": Object {
    "unit": "PERCENT",
    "value": 140,
  },
}
`;

exports[`transformTextStyleProperties lineHeight auto 1`] = `
Object {
  "lineHeight": Object {
    "unit": "AUTO",
  },
}
`;

exports[`transformTextStyleProperties text shadow 1`] = `
Object {
  "effects": Array [
    Object {
      "blendMode": "NORMAL",
      "color": Object {
        "a": 1,
        "b": 0,
        "g": 0,
        "r": 0,
      },
      "offset": Object {
        "x": 10,
        "y": 20,
      },
      "radius": 15,
      "type": "DROP_SHADOW",
      "visible": true,
    },
  ],
}
`;

exports[`transformTextStyleProperties textAlign auto 1`] = `Object {}`;

exports[`transformTextStyleProperties textAlign center 1`] = `
Object {
  "textAlignHorizontal": "CENTER",
}
`;

exports[`transformTextStyleProperties textAlign justify 1`] = `
Object {
  "textAlignHorizontal": "JUSTIFIED",
}
`;

exports[`transformTextStyleProperties textDecorationLine line-through 1`] = `
Object {
  "textDecorationLine": "STRIKETHROUGH",
}
`;
